### PR TITLE
Bug fixes & additional functionality

### PR DESCRIPTION
**Pull request by VakarisZ**

### CODE REVIEW

**Comment by danielguardicore** [comment]
- Monkey deployment dir should either be where the script is run or a command line parameter. Preferably both. 

### Windows 
- In Powershell, the functionality should be wrapped inside a function. This way we get some Usage for free.

- I wouldn't install any program that changes env for the user. Not git, not Python or the Python VC compiler. Ask the user to download and provide URL. 

- What about the sambacry binaries? Do we need to make them available? That's also required if we're already downloading mimikatz.

- When extract the mongo folder, why do we need to move the items rather than just extract to the proper place?

### Linux

We should minimise usage of sudo. This means pip install --user (if possible),avoiding sudo apt update. etc. Where required, lets do it. Also, lets check for sudo right up front instead of surprising the user with a sudo prompt in the middle.
https://stackoverflow.com/questions/18215973/how-to-check-if-running-as-root-in-a-bash-script

You can use uname `-m` rather than `-mrs`, easier checking.

**Comment by VakarisZ** [comment]
Windows.
2. How come? This script doesn't automatically install it just downloads and launches the installer, so if a person doesn't want to install something he can click cancel.
3. As I've heard you can't build linux binaries on windows, so what's the point in building samba? Also, only bash script for building samba binaries is available, so I presumed it's not meant for windows.
4. Required mongo files are extracted with a wrapper folder that also contains unnecessary files. Because file sizes are rather small I didn't waste time trying to find how to extract only particular files and strip the wrapper folder. I just move the the files we need and delete everything else after extraction.

**Comment by danielguardicore** [comment]
2 - True, we can keep it that way. Generally installers have a -Quiet option. This would mean in our case, that we'd pass in automatic installs for all our subcomponents. For example, in Python you can auto install using the following flags.
https://www.python.org/download/releases/2.4/msi/
https://docs.python.org/3/using/windows.html

3 - We can't build (easily) linux code on Windows, without using WSL. But we should make sure the python works and it relies on binaries. Two options (a) offer them to download from our github (b) build using WSL.
I vote for (a) or nothing.

4- ACK.

File: deploy_linux.sh [hunk 1]
```diff
@@ -1,38 +1,59 @@
...
-    git clone ${MONKEY_GIT_URL} ${MONKEY_HOME_DIR} 2>&1 || handle_error
+if [[ ! -d "$monkey_home/monkey" ]]; then # If not already cloned
+    git clone ${MONKEY_GIT_URL} ${monkey_home} 2>&1 || handle_error
+    chmod 774 -R ${monkey_home}
 fi
 
 # Create folders
-log_message "Creating island dirs under $MAIN_ISLAND_PATH"
+log_message "Creating island dirs under $ISLAND_PATH"
 mkdir -p ${MONGO_BIN_PATH}
```
Comment (danielguardicore) [inline deploy_linux.sh:55]: Why not check the rest for errors?

File: deploy_linux.sh [hunk 2]
```diff
@@ -41,8 +62,8 @@ if [[ ${python_version} == *"command not found"* ]] || [[ ${python_version} != *
 fi
 
 log_message "Installing island requirements"
-requirements="$MAIN_ISLAND_PATH/requirements.txt"
-sudo python -m pip install -r ${requirements} || handle_error
+requirements="$ISLAND_PATH/requirements.txt"
+python -m pip install -r ${requirements} || handle_error
```
Comment (danielguardicore) [inline deploy_linux.sh]: This won't work. Add a --user flag. Also avoids clobbering a systems env.

File: deploy_windows.ps1 [hunk 3]
```diff
@@ -1,202 +1,208 @@
...
-}
-$mk64_path = Join-Path -Path $binDir -ChildPath $MK64_DLL
-if(!(Test-Path -Path $mk64_path )){
-    "Downloading mimikatz 32 binary"
-    $webClient.DownloadFile($MK64_DLL_URL, $mk64_path)
-}
+    # We check if python is installed
+    try
+    {
+        $version = cmd.exe /c '"python" --version  2>&1'
```
Comment (danielguardicore) [inline deploy_windows.ps1:46]: You can use `Get-Command` :)

File: deploy_windows.ps1 [hunk 4]
```diff
@@ -1,202 +1,208 @@
...
+    # Install requirements
+    $reqPath = Join-Path -Path $monkey_home -ChildPath $MONKEY_ISLAND_DIR | Join-Path -ChildPath "\requirements.txt" -ErrorAction Stopd
+    "Upgrading pip..."
+    $output = cmd.exe /c 'python -m pip install --upgrade pip 2>&1'
+    $output
+    if ( $output -like '*No module named pip*' ) {
+        "Make sure pip module is installed and re-run this script."
+        return
+    }
+    & python -m pip install -r $reqPath
```
Comment (danielguardicore) [inline deploy_windows.ps1]: `--user`

File: deploy_windows.ps1 [hunk 5]
```diff
@@ -1,202 +1,208 @@
...
+    # Install requirements from monkey to be able to develop monkey itself
+    $reqPath = Join-Path -Path $monkey_home -ChildPath $MONKEY_DIR | Join-Path -ChildPath "\requirements.txt"
```
Comment (danielguardicore) [inline deploy_windows.ps1]: Just define a different variable...

File: deploy_windows.ps1 (hunk 3 above)
Comment (danielguardicore) [reply deploy_windows.ps1:46]: You even use it later on

File: deploy_windows.ps1 [hunk 6]
```diff
@@ -1,202 +1,208 @@
...
+    & python -m pip install -r $reqPath
+
+    # Transfer python file to local directory
+    "Copying python folder to bin"
+    Copy-Item $PYTHON_PATH -Destination (Join-Path -Path $binDir -ChildPath "Python27") -Recurse -ErrorAction SilentlyContinue
```
Comment (danielguardicore) [inline deploy_windows.ps1]: Why do this for this env if it's for dev. This can easily confuse people adding requirements.

Comment (danielguardicore) [reply deploy_windows.ps1]: If we need this in terms of paths, I'd suggest adding a symlink and not adding a lot of stuff to the drive.

File: deploy_linux.sh (hunk 1 above)
Comment (VakarisZ) [reply deploy_linux.sh:55]: Can't think of any situation where only one of them occur (and all errors require the same fix). Thus we want user to know everything he needs to fix at once (all the folders that needs permissions for e.g.).

File: deploy_windows.ps1 (hunk 3 above)
Comment (VakarisZ) [reply deploy_windows.ps1:46]: I can, but not in this place. Get-Command doesn't output results of a command. What if 'python' command is used for python3 ? In that case we still need to download version 2.7. Also, more readable.

**Pull request merged**